physics/GFS_surface_composites.F90

```
@@ -348,46 +374,56 @@ end subroutine GFS_surface_composites_post_finalize
         fm10_wat, fm10_lnd, fm10_ice, fh2_wat, fh2_lnd, fh2_ice, tsurf_wat, tsurf_lnd, tsurf_ice, cmm_wat, cmm_lnd, cmm_ice,    &
         chh_wat, chh_lnd, chh_ice, gflx_wat, gflx_lnd, gflx_ice, ep1d_wat, ep1d_lnd, ep1d_ice, weasd_wat, weasd_lnd, weasd_ice, &
         snowd_wat, snowd_lnd, snowd_ice,tprcp_wat, tprcp_lnd, tprcp_ice, evap_wat, evap_lnd, evap_ice, hflx_wat, hflx_lnd,      &
-        hflx_ice, qss_wat, qss_lnd, qss_ice, tsfc_wat, tsfc_lnd, tsfc_ice
+        hflx_ice, qss_wat, qss_lnd, qss_ice, tsfc_wat, tsfc_lnd, tsfc_ice, zorlo, zorll, zorli
 
-      real(kind=kind_phys), dimension(:),   intent(inout) :: zorl, zorlo, zorll, zorli, cd, cdq, rb, stress, ffmm, ffhh, uustar, fm10, &
-        fh2, tsurf, cmm, chh, gflx, ep1d, weasd, snowd, tprcp, evap, hflx, qss, tsfc, tsfco, tsfcl, tisfc
+      real(kind=kind_phys), dimension(:),   intent(inout) :: zorl, cd, cdq, rb, stress, ffmm, ffhh, uustar, fm10,               &
+        fh2, cmm, chh, gflx, ep1d, weasd, snowd, tprcp, evap, hflx, qss, tsfc, tsfco, tsfcl, tisfc
 
       real(kind=kind_phys), dimension(:),   intent(in   ) :: tice ! interstitial sea ice temperature
       real(kind=kind_phys), dimension(:),   intent(inout) :: hice, cice
       real(kind=kind_phys),                 intent(in   ) :: min_seaice
+      real(kind=kind_phys),                 intent(in   ) :: rd, rvrdm1
 
       real(kind=kind_phys), dimension(:,:), intent(in   ) :: tiice
       real(kind=kind_phys), dimension(:,:), intent(inout) :: stc
 
+      ! Additional data needed for calling "stability"
+      logical,                            intent(in   ) :: thsfc_loc
+      real(kind=kind_phys),               intent(in   ) :: grav
+      real(kind=kind_phys), dimension(:), intent(in   ) :: prsik1, prslk1, prslki, z1
+      real(kind=kind_phys), dimension(:), intent(in   ) :: ztmax_wat, ztmax_lnd, ztmax_ice
```

@tanyasmirnova @benwgreen @joeolson42 @grantfirl There is one flaw when not using `sfc_diff` but, for example MYNN sfc or GFDL sfc when the fractional landmask is used. The arrays `ztmax_wat`, `ztmax_lnd`, `ztmax_ice` are not calculated in either of those two schemes, only in `sfc_diff.f`. They will be zero and taking the log of either of these in line 493 will trigger floating point exceptions.

Right now, we do not have regression tests with fractional landmask using either of the two alternative surface layer schemes, but we should fix this as part to this PR.

---

In `sfc_diff.f`, I put in Lines 180-183 code to initialize `ztmax_wat`, `ztmax_lnd`, `ztmax_ice` to values of 1.0, to avoid taking log(0). So I guess those lines could be moved somewhere else to that when MYNN/GFDL sfc is used, those arrays will still be initialized to values of 1.0?

---

I would initialize and compute those arrays correctly in each of the surface layer schemes. Leaving it uninitialized or initialized to zero in `GFS_typedefs.F90` means that the code will tell us if a surface layer scheme doesn't compute the necessary quantities, rather than silently using a default value.